hooks/useAvalancheForecast.ts

```diff
@@ -1,57 +1,130 @@
 import React from 'react';
 
-import {QueryClient, useQuery} from '@tanstack/react-query';
+import {QueryClient, useQuery, useQueryClient} from '@tanstack/react-query';
 import axios, {AxiosError} from 'axios';
 
 import * as Sentry from 'sentry-expo';
 
 import Log from 'network/log';
 
 import {ClientContext, ClientProps} from 'clientContext';
-import {Product, productSchema} from 'types/nationalAvalancheCenter';
+import {logQueryKey} from 'hooks/logger';
+import AvalancheForecastByID from 'hooks/useAvalancheForecastById';
+import AvalancheForecastFragment from 'hooks/useAvalancheForecastFragment';
+import {AvalancheCenter, AvalancheCenterID, Product, productSchema} from 'types/nationalAvalancheCenter';
+import {nominalForecastDate, nominalForecastDateString, RequestedTime} from 'utils/date';
 import {ZodError} from 'zod';
 
-export const useAvalancheForecast = (fragment: Product) => {
+export const useAvalancheForecast = (center_id: AvalancheCenterID, center: AvalancheCenter, zone_id: number, requestedTime: RequestedTime) => {
+  const queryClient = useQueryClient();
   const {nationalAvalancheCenterHost} = React.useContext<ClientProps>(ClientContext);
-  const forecastId = fragment?.id;
+
+  const expiryTimeHours = center?.config.expires_time;
+  const expiryTimeZone = center?.timezone;
 
   return useQuery<Product, AxiosError | ZodError>({
-    queryKey: queryKey(nationalAvalancheCenterHost, forecastId),
-    queryFn: () => fetchProduct(nationalAvalancheCenterHost, forecastId),
-    enabled: !!forecastId,
-    staleTime: 60 * 60 * 1000, // re-fetch in the background once an hour (in milliseconds)
+    queryKey: queryKey(nationalAvalancheCenterHost, center_id, zone_id, requestedTime, expiryTimeZone, expiryTimeHours),
+    queryFn: async () => fetchAvalancheForecast(queryClient, nationalAvalancheCenterHost, center_id, zone_id, requestedTime, expiryTimeZone, expiryTimeHours),
+    enabled: !!expiryTimeHours,
     cacheTime: 24 * 60 * 60 * 1000, // hold on to this cached data for a day (in milliseconds)
   });
 };
 
-function queryKey(nationalAvalancheCenterHost: string, forecastId: number) {
-  return ['avalanche-forecast', {host: nationalAvalancheCenterHost, forecast: forecastId}];
+// the query we make to the NAC API does not make use of the nominal date for which we're asking, but we want a good
+// user experience around the cut-over moment between nominal dates. Consider a user that has loaded this query before
+// the cut-over, and has a cached version of the previous day's forecasts. When they refresh their view after the
+// cut-over, we want to render only once we've done a fresh load, since we expect to have new data, instead of using
+// the cache. By adding the nominal date to the query key, we get this behavior.
```

Comment:
> we want to render only once we've done a fresh load

I disagree pretty strongly with this statement. I would much rather see new data pop in on that rare occasion (we could make the transition smoother if it happens), in exchange for a much faster browsing experience. 

Also I think this behaves poorly if I'm offline just after midnight - there might be data in cache that's not very stale but I can't see it because it was cached as "yesterday's news". Whereas if it was cached under the "latest" key then I'd be able to see it, even if it was slightly stale.

Reply:
> I disagree pretty strongly with this statement. I would much rather see new data pop in on that rare occasion 

The implication for this app, though, is that we show the wrong forecast to the user and can't be certain that we ever re-fresh and show the correct one. At first blush that seems like something we can't afford to do, even if, as you point out, the way that an application would handle this in the usual case is to prefer a faster experience.

> Also I think this behaves poorly if I'm offline just after midnight - there might be data in cache that's not very stale but I can't see it because it was cached as "yesterday's news". Whereas if it was cached under the "latest" key then I'd be able to see it, even if it was slightly stale.

We use a nominal forecast date, which cuts over at e.g. 6pm for NWAC. Midnight or not should have absolutely no bearing on this.